linera-execution/src/lib.rs

```diff
@@ -1074,6 +1076,53 @@ impl std::fmt::Debug for Bytecode {
     }
 }
 
+/// A compressed WebAssembly module's bytecode.
+#[derive(Clone, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct CompressedBytecode {
+    #[serde(with = "serde_bytes")]
+    compressed_bytes: Vec<u8>,
+}
+
+impl From<&Bytecode> for CompressedBytecode {
+    fn from(bytecode: &Bytecode) -> Self {
+        let compressed_bytes = zstd::stream::encode_all(&*bytecode.bytes, 19)
+            .expect("Compressing bytes in memory should not fail");
+
+        CompressedBytecode { compressed_bytes }
+    }
+}
+
+impl From<Bytecode> for CompressedBytecode {
+    fn from(bytecode: Bytecode) -> Self {
+        CompressedBytecode::from(&bytecode)
+    }
+}
+
+impl TryFrom<&CompressedBytecode> for Bytecode {
+    type Error = ExecutionError;
+
+    fn try_from(compressed_bytecode: &CompressedBytecode) -> Result<Self, Self::Error> {
+        let bytes = zstd::stream::decode_all(&*compressed_bytecode.compressed_bytes)
+            .map_err(ExecutionError::InvalidCompressedBytecode)?;
+
+        Ok(Bytecode { bytes })
+    }
+}
+
+impl TryFrom<CompressedBytecode> for Bytecode {
+    type Error = ExecutionError;
+
+    fn try_from(compressed_bytecode: CompressedBytecode) -> Result<Self, Self::Error> {
+        Bytecode::try_from(&compressed_bytecode)
+    }
+}
+
+impl std::fmt::Debug for CompressedBytecode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        f.debug_tuple("CompressedBytecode").finish()
```

Comment:
`finish_non_exhaustive`?

Reply:
Unfortunately, that's not available for [tuples](https://doc.rust-lang.org/std/fmt/struct.DebugTuple.html) :(

Reply:
(Ah, right! We coudl just explicitly write `"CompressedBytecode(..)"` then. Not sure if that's better. Your call!)

Reply:
Wait. Even better: it's not actually a tuple :man_facepalming: I fixed it and added the `non_exhaustive`. Sorry I missed this :sweat: